### ISSUE TITLE
Parallel image pull should pull unique images only once

### ISSUE THREAD

**Issue by DocX** [kubernetes/kubernetes#125164]
### What would you like to be added?

Creating large amount of pods that all use the same image causes the kubelet to start pull per each pod, when parallel pull is enabled. This causes the reigstry QPS to be used, slowing down other pods with other images and also wastes the bandwidth for downloading the same images multiple times.

If same image is being already downloaded in parallel, other pods should just wait for it, not starting new download and not using up registry QPS. 

### Why is this needed?

We have use case where we start large number (500+) pods at the same time where many pods gets scheduled on single node (100+), and most pods have same init container with the same image.

This causes lot of bandwidth on the node to be used during the startup of those pods, and also some pods are delayed due to `ImagePullBackoff` caused by `pull QPS exceeded`.

**Comment by saschagrunert** [comment]
The question would be if we should do that within the container runtime or the kubelet itself. I see pros and cons for both sides. :thinking:

**Comment by saschagrunert** [comment]
@kubernetes/sig-node-bugs we could add a broadcast channel per image to be pulled o the parallel image puller implementation: 
https://github.com/kubernetes/kubernetes/blob/1d5589e4910ed859a69b3e57c25cbbd3439cd65f/pkg/kubelet/images/puller.go#L42-L45

Then we check if a pull is already in progress and use the broadcast instead of a new pull in 
https://github.com/kubernetes/kubernetes/blob/1d5589e4910ed859a69b3e57c25cbbd3439cd65f/pkg/kubelet/images/puller.go#L54-L74

Do you have any thoughts on that?

**Comment by sftim** [comment]
@saschagrunert is this a bug report or a feature request?

**Comment by saschagrunert** [comment]
@sftim I think it's more like fixing a bug from a corner-case.

**Comment by DocX** [comment]
I want to just share some more thoughts I have about the usability / transparency of this behaviour. I think to make it not confusing and be more "predictable", it should behave like this:

* If `imagePullPolicy: Always` then no deduplication should happen, even in the case the image is specified with the image digest. 
* If `imagePullPolicy: IfNotPresent` then if the image with the same URL (regardless digest or tag) is being pulled in other thread already, it should wait for it. This behaviour is consistent with no deduplication/waiting, since the "presence" of the image is based on the URL and not the specific image digest (i.e. if the image has been updated in meanwhile in registry under the same tag)

Does it make sense?

**Comment by aojea** [comment]
Another related an interesting observation that will help to improve pod startup is to have affinity to pull images from the same pod, I've seen this in an internal cluster today.

If a Pod has 3 containers per example, it will not start until those 3 images are available, right now the puller try to get images randomly, so if the pod is unlucky and the first 2 containers images are pulled but the last is at the end of the queue, the pod startup will be determined by this last image.

Scenario: 
node start and 100 pods are scheduled, each pod has 3 containers, if in the worst case, pod1 pulls container1, and container2 image, but then container3 is pulled after pod1-99 are processed, pod1 startup will be the sum of all the pods

A priority queue in the poller to prioritize images of the same pod could be used to solve this problem, since the user will prefer to pods to be running faster (cc @dchen1107 @yujuhong )

**Comment by saschagrunert** [comment]
> puller try to get images randomly

Do you mind pointing me to the code where this happens? I see that we call `pullImage` for each container in a pod via `EnsureImageExists` in:

https://github.com/kubernetes/kubernetes/blob/a716ea756d87f60900dbbb500fc27ae30f7bd384/pkg/kubelet/kuberuntime/kuberuntime_container.go#L210

The puller can be either sequential or parallel, in case it's a parallel puller it uses a channel which should pull also in order ([with a max concurrency of 10](https://github.com/kubernetes/kubernetes/blob/a716ea756d87f60900dbbb500fc27ae30f7bd384/pkg/kubelet/images/puller.go#L77C7-L77C27)), because of the loop:

https://github.com/kubernetes/kubernetes/blob/a716ea756d87f60900dbbb500fc27ae30f7bd384/pkg/kubelet/images/puller.go#L109

What am I missing?

**Comment by SergeyKanzhelev** [comment]
> A priority queue in the poller to prioritize images of the same pod could be used to solve this problem, since the user will prefer to pods to be running faster (cc [@dchen1107](https://github.com/dchen1107) [@yujuhong](https://github.com/yujuhong) 

This is not as straightforward as it sounds. Will user prefer to pull all init container's images for other 99 pods so it can start executing init containers for other 99 images and then start the first one? Ultimately it may be faster overall. So priority queue needs to be very carefully balanced. There is also matter of Pods priority, which is similarly vague. Some users will want all system pods to be ready before running user pod (prioritize system pods image pulls). Others are happy to postpone some metrics collection Pod - it is critical for them to start running their Pod asap (prioritize user pod images).

**Comment by SergeyKanzhelev** [comment]
I believe the original issue is about the fact that kubelet doesn't know how to dedup images. Since images that are defined identically may theoretically result in different digests, kubelet effectively fills up all slots in parallel image pulling with the "same" image - waiting on a single image pull by runtime. I would think the same effect can be seen if different images have the same huge base layer, two image pulls will effectively block two "slots" of parallel image pulling on a single layer pull.

Fixing this will be a feature and I do not think we should block the KEP https://github.com/kubernetes/enhancements/issues/3673 on this feature. KEP is improving things This suggestion is another step towards the improvement in smart image pulling.
